capture.js

```diff
@@ -0,0 +1,53 @@
+const capture = {
+
+  init() {
+    this.addListeners();
+    this.prevSite = null;
+    this.nextSite = null;
+  },
+
+  addListeners() {
+    // listen for each HTTP response
+    browser.webRequest.onResponseStarted.addListener(this.collectRequests, {urls: ['<all_urls>']}
+    );
+    // listen for when user visits a new site
+    browser.tabs.onUpdated.addListener(this.checkNewSite);
+  },
+
+  async collectRequests(response) {
+    const tab = await browser.tabs.get(response.tabId);
+    let tabUrl = new URL(tab.url);
+    tabUrl = tabUrl.hostname;
+    let targetUrl = new URL(response.url);
+    targetUrl = targetUrl.hostname;
+    let originUrl = new URL(response.originUrl);
+    originUrl = originUrl.hostname;
+
+    // capture third party requests
+    if (targetUrl !== originUrl) {
+      const thirdPartyData = {
+        document: tabUrl,
+        target: targetUrl,
+        origin: originUrl,
+        requestTime: response.timeStamp
+      };
+      console.log('storage.thirdPartyRequest:', tabUrl, thirdPartyData);
+    }
+  },
+
+  checkNewSite(tabId, changeInfo, tab) {
+    if (changeInfo.url) {
+      this.nextSite = changeInfo.url;
+    }
+    //favicon not available until tab status is complete; opening a new tab (about:blank) does not trigger a call to storage API
+    if ((this.nextSite !== this.prevSite) && (changeInfo.status === 'complete') && (this.nextSite !== null) && (this.nextSite !== 'about:blank')) {
```

Comment:
The first call a tab makes isn't ignored. I have tested it. This is thanks to `this.nextSite` and `this.prevSite`. I initialize `this.prevSite` to `null` and when the tab updates and status is complete, that differs from the tab's new URL and it calls the storage API.
The `this.nextSite` and `this.prevSite` are the way I know whether the request represents a new first party request (so I don't make a bunch of calls to storage needlessly). This won't prevent complete redundancy, but it helps a lot. It sounds like you prefer storage to handle that. Why is that better? I'm perfectly okay changing it, just want to make sure I am following the reasoning.